shared/utility/math/filter/eigen/UKF.h

```diff
@@ -0,0 +1,352 @@
+        class UKF {
+
+        public:
+            // The model
+            using Model = FilterModel<Scalar>;
+            Model model;
+
+            // Dimension types for vectors and square matrices
+            using StateVec = Eigen::Matrix<Scalar, Model::size, 1>;
+            using StateMat = Eigen::Matrix<Scalar, Model::size, Model::size>;
+
+        private:
+            // The number of sigma points
+            static constexpr unsigned int NUM_SIGMA_POINTS = (Model::size * 2) + 1;
+
+            using SigmaVec       = Eigen::Matrix<Scalar, NUM_SIGMA_POINTS, 1>;
+            using SigmaMat       = Eigen::Matrix<Scalar, Model::size, NUM_SIGMA_POINTS>;
+            using SigmaSquareMat = Eigen::Matrix<Scalar, NUM_SIGMA_POINTS, NUM_SIGMA_POINTS>;
+
+            // Our estimate and covariance
+            StateVec mean;
+            StateMat covariance;
+
+            // Our sigma points for UKF
+            StateVec sigma_mean;
+            SigmaMat sigma_points;
+
+            SigmaMat centred_sigma_points;  // X in Steves kalman theory
+            SigmaVec d;
+            SigmaSquareMat covariance_update;  // C in Steves kalman theory
+
+            // The mean and covariance weights
+            SigmaVec mean_weights;
+            SigmaVec covariance_weights;
+
+        private:
+            // UKF variables
+            Scalar covariance_sigma_weight;
+
+            /**
+             * @brief Generate new sigma points given our mean and covariance.
+             */
+            template <typename T, int S>
+            static Eigen::Matrix<T, S, NUM_SIGMA_POINTS> generate_sigma_points(const Eigen::Matrix<T, S, 1>& mean,
+                                                                               const Eigen::Matrix<T, S, S>& covariance,
+                                                                               const T& sigma_weight) {
+
+                Eigen::Matrix<T, S, NUM_SIGMA_POINTS> points;
+
+                // Our first row is always the mean
+                points.col(0) = mean;
+
+                // Get our Cholesky decomposition
+                Eigen::LLT<StateMat> cholesky(sigma_weight * covariance);
+                if (cholesky.info() == Eigen::Success) {
+                    // Put our values in either end of the matrix
+                    StateMat chol = cholesky.matrixL().toDenseMatrix();
+                    for (unsigned int i = 1; i < Model::size + 1; ++i) {
+                        points.col(i)               = (mean + chol.col(i - 1));
+                        points.col(i + Model::size) = (mean - chol.col(i - 1));
+                    }
+                }
+                else {
+                    throw std::runtime_error("Cholesky decomposition failed");
+                }
+
+                return points;
+            }
+
+            /**
+             * @brief Calculate the the mean given a set of sigma points.
+             *           This version
```

Comment:
This version.... ?

Reply:
Did I add this? Or is that in your version too?

Reply:
Apparently I started writing something and forgot, It's in mine too

Reply:
No idea might be something that went away with Eigen